### DISCUSSION TITLE
DAI 2024-2025 - Practical work 3 - Pedro Alves da Silva & Gonçalo Carvalheiro Heleno

### DISCUSSION THREAD

**Discussion by PedroAS7** [heig-vd-dai-course/heig-vd-dai-course#538] [Show and tell]
For this last practical work, we'll be creating an HTTP server that will serve as a sort of GPG public key directory. The idea is that users might have multiple GPG key pairs, and we'd like our web server to store all of the public keys to allow their easy retrieval. A real-world usage scenario would be querying our web server to retrieve a public key to encrypt an email we'd want to send to the owner of that key pair.

The web server would be able to handle GET, POST, UPDATE, and DELETE methods for GPG public keys, as well as GET, POST, and UPDATE methods to handle the key owners. Depending on how it goes, we might add more resources.

**Comment by GeraudSilvestri** [comment]
Good for me

**Comment by lentidas** [comment]
@heig-vd-dai-course/teaching-staff

We have a question an a few of the grading criteria:

> 15 - The README (or repository) contains instructions how to install and configure the virtual machine with each step

We have created the VM with Terraform and Ansible (see code [here](https://github.com/lentidas/DAI-2425-PW3/tree/main/infra)). So, do we need to provide instructions on how to create and configure the VM or providing instructions on how to run the Terraform/Ansible code is enough?

> 16 - The README (or repository) contains explains how to configure the DNS zone to access your web application

Microsoft Azure provides a default DNS domain for public IPs (see [here](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/public-ip-addresses#domain-name-label) and [here](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/public_ip#domain_name_label-1)). We thought that using it was good enough. Do you want us to configure a DNS zone with a proper domain?

> 22 - The requests are cached to improve performance until the data changes

The caching needs to be one of the HTTP caching methods or can it be any other form of caching implemented on the Java application? We pose this question because we are using an SQL database for data persistency, so we thought that caching some of the data on-memory in the Java application was good enough cache. However we started asking ourselves if that was not what was asked of us...

Thank you very much 🙂

**Reply by ludelafo** [reply]
Hello @lentidas,

Thank you for your message.

> We have created the VM with Terraform and Ansible (see code [here](https://github.com/lentidas/DAI-2425-PW3/tree/main/infra)). So, do we need to provide instructions on how to create and configure the VM or providing instructions on how to run the Terraform/Ansible code is enough?

In your case, instructions to run the Terraform/Ansible code is enough. What we aim to see is how you got/set up the infrastructure you have so other people could understand/replicate your environment in which your web application runs.

> Microsoft Azure provides a default DNS domain for public IPs (see [here](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/public-ip-addresses#domain-name-label) and [here](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/public_ip#domain_name_label-1)). We thought that using it was good enough. Do you want us to configure a DNS zone with a proper domain?

I wasn't aware about this feature, thanks for letting me know! It is good enough indeed.

> The caching needs to be one of the HTTP caching methods or can it be any other form of caching implemented on the Java application? We pose this question because we are using an SQL database for data persistency, so we thought that caching some of the data on-memory in the Java application was good enough cache. However we started asking ourselves if that was not what was asked of us...

We are expected to see the caching mechanisms seen in the [Caching and performance](https://github.com/heig-vd-dai-course/heig-vd-dai-course/tree/main/23-caching-and-performance) chapter (a status code `200 OK` on the first request and a conditional `302 Not Modified` on a future request if the data hasn't changed or another `200 OK` if the data has been updated since the last time/version you got). It can be implemented in many ways and we have provided one way in the course material to do it but the conditional caching mechanism must be seen and can be tested with curl and/or the browser.

However, I'm not sure if I understood your last point correctly, feel free to let me know if I haven't fully answered your question.

Please let me know if you have any more questions, or if I have answered them all.

Have a nice day!

**Reply by lentidas** [reply]
You answered all my questions. Thank you! We were not implementing the caching like you expected at all, I'm glad I asked 😄

Have a nice day!

**Comment by lentidas** [comment]
https://github.com/lentidas/DAI-2425-PW3/commit/28010aa5355a2e3913046f45d6cfecb4b0a42c32
